File: src/Assets/Scripts/NetworkManager.js
#pragma strict
#pragma implicit
#pragma downcast

private var typeName = "SpectrumMP";
private var gameName = "Untitled";
private var hostList = [];
private var player : GameObject;
private var joinIP : String = "localhost";
private var chatList = Array();
var userText : String = "";
var playerPrefab : GameObject;
var showUI = true;

function Start() {
	MasterServer.ipAddress = "121.214.63.194";
	MasterServer.port = 23466;
	Network.natFacilitatorIP = "121.214.63.194";
	Network.natFacilitatorPort = 23466;
}

function SpawnPlayer() {
	player = Network.Instantiate(playerPrefab, Vector3(0, 20, 0), Quaternion.identity, 0);
}

function StartServer () {
	Network.InitializeServer(4, 62010, !Network.HavePublicAddress());
	MasterServer.RegisterHost(typeName, gameName);
}

function OnServerInitialized() {
	SpawnPlayer();
	SendGameMessage("Game hosted on port 62010");
}

function OnGUI() {
	if (showUI) {
		if (!Network.isClient && !Network.isServer) {
			gameName = GUI.TextField(Rect(200, 100, 100, 20), gameName, 20);
			joinIP = GUI.TextField(Rect(200, 130, 100, 20), joinIP, 20);
			if (GUI.Button(Rect(100, 100, 100, 20), "Open Server")) {
				SendGameMessage("Creating server '" + gameName + "'...");
				StartServer();
			}
			
			if (GUI.Button(Rect(100, 160, 200, 20), "Refresh Hosts")) {
				SendGameMessage("Refreshing hosts...");
				RefreshHostList();
			}
			
			if (GUI.Button(Rect(100, 130, 100, 20), "Join IP")) {
				SendGameMessage("Joining IP " + joinIP + "...");
				Network.Connect(joinIP, 62010);
			}
			
			if (hostList.length >= 0) {
				for (var i = 0; i < hostList.length; i++) {
					var hData : HostData = hostList[i];
					if (GUI.Button(Rect(100, 190 + (30 * i), 200, 20), hData.gameName + " | " + hData.connectedPlayers + "/4")) {
						SendGameMessage("Joining host " + hData.gameName + "...");
						JoinServer(hData);
					}
				} 
			}
		} else if (Network.isServer) {
			if (GUI.Button(Rect(0, 45, 100, 20), "Close server")) {
				Network.Disconnect();
			}
			GUI.Box(Rect(0, 70, 100, 20), "Kick players");
			for (i = 0; i < Network.connections.Length; i++) {
				if (GUI.Button(Rect(0, 95 + 25*i, 100, 20), Network.connections[i].ipAddress)) {
					Network.CloseConnection(Network.connections[i], true);
				}
			}
		} else {
			if (GUI.Button(Rect(0, 45, 100, 20), "Disconnect")) {
				Network.Disconnect();
			}
		}
		GUI.Box(Rect(0, 20, 100, 20), "Username");
		
		for (i = 0; i < chatList.length; i++) {
			j = chatList.length - (i+1);
			var message : String = chatList[j];
			GUI.Label(Rect(0, Screen.height - (25 * (i+1)), 300, 25), message);
		}
	}
}

function RefreshHostList() {
	MasterServer.RequestHostList(typeName);
}

function OnMasterServerEvent(msEvent : MasterServerEvent) {
	if (msEvent == MasterServerEvent.HostListReceived) {
		hostList = MasterServer.PollHostList();
	}
}

function JoinServer(hostData : HostData) {
	Network.Connect(hostData);
}

function OnConnectedToServer() {
	SpawnPlayer();
}

function OnPlayerConnected(player: NetworkPlayer) {
	SendGameMessage("Player joined the game." + NumberPlayers(true));
}

function OnPlayerDisconnected(player: NetworkPlayer) {
	Network.RemoveRPCs(player);
	Network.DestroyPlayerObjects(player);
	SendGameMessage("Player left the game." + NumberPlayers(false));
}

function OnDisconnectedFromServer(info : NetworkDisconnection) {
	if (Network.isServer) {
		SendGameMessage("Local server connection disconnected");
	} else {
		if (info == NetworkDisconnection.LostConnection) {
			SendGameMessage("Lost connection to the server.");
		} else {
			SendGameMessage("Disconnected from the server.");
		}
	}
}

function OnFailedToConnect(error: NetworkConnectionError) {
	SendGameMessage("Could not connect to server:");
	SendGameMessage(error.ToString());
}
	
function NumberPlayers(add : boolean) {
	var plusOne : int = 0;
	if (add) { plusOne = 1; }
	var str : String = (Network.connections.Length + plusOne + "/" + Network.maxConnections);
	return str;
}

function SendGameMessage(msg : String) {
	chatList.push(msg);
	Debug.Log(FormatTime(Time.time) + " " + msg);
	if (chatList.length > 6) {
		chatList.shift();
	}
}

function FormatTime(time : float) {
	var intTime : int = time;
	var minutes : int = intTime / 60;
	var seconds : int = intTime % 60;
	var fraction : int = time * 10;
	fraction = fraction % 10;
	
	var timeText = minutes.ToString("00") + ":";
	timeText += seconds.ToString("00") + ":";
	timeText += fraction.ToString("00");
	return timeText;
}

File: src/Assets/Scripts/NetworkManager.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class NetworkManager : MonoBehaviour {

	string typeName = "SpectrumMP";
	string joinIP = "localhost";
	public HostData[] hostList = {};
	List<string> chatList = new List<string> ();
	GameManager gameManager;
	
	public string displayName = "";
	public GameObject playerPF;
	public bool showUI = true;
	public int hostPort = 62010;
	
	//public GUIStyle style;
	public GUIStyle chatStyle;
	
	// MonoBehaviour
	void Start () {
		MasterServer.ipAddress = "121.214.63.194";
		MasterServer.port = 23466;
		Network.natFacilitatorIP = "121.214.63.194";
		Network.natFacilitatorPort = 23466;
		gameManager = (GameManager)GameObject.FindGameObjectWithTag ("GameController").GetComponent<GameManager> ();
	}

	Object SpawnPlayer () {
		return Network.Instantiate (playerPF, new Vector3 (0, 20, 0),
									Quaternion.identity, 0);
	}
	
	public void StartServer (string gameName) {
		Network.InitializeServer (4, hostPort, !Network.HavePublicAddress ());
		MasterServer.RegisterHost (typeName, gameName);
	}
	
	public void StartPrivate () {
	}
	
	public void ConnectTo (string address, int port) {
		Network.Connect (address, port);
	}
	
	// MonoBehaviour
	void OnServerInitialized () {
		SpawnPlayer ();
		SendGameMessage ("Game hosted on port " + hostPort);
	}
	
	// MonoBehaviour
	void OnGUI () {
		if (showUI) {
			if (Network.isServer) {
				if (GUI.Button (new Rect (10, 55, 100, 20), "Close server")) {
					Network.Disconnect ();
				}
				GUI.Box (new Rect (10, 80, 100, 20), "Kick players");
				for (int i = 0; i < Network.connections.Length; i++) {
					if (GUI.Button (new Rect(10, 105 + 25*i, 100, 20), Network.connections[i].ipAddress)) {
						Network.CloseConnection (Network.connections[i], true);
					}
				}
			} else {
				if (GUI.Button (new Rect (10, 55, 100, 20), "Disconnect")) {
					Network.Disconnect ();
				}
			}
			
			GUI.Box( new Rect (10, 30, 100, 20), "Username");
			
			for (int i = 0; i < chatList.Count; i++) {
				int j = chatList.Count - (i + 1);
				string message = chatList[i];
				GUI.Label (new Rect (10, Screen.height - (25 * (j + 1)) - 10, 300, 25), message, chatStyle);
			}
		}
	}
	
	public void RefreshHostList () {
		MasterServer.RequestHostList (typeName);
	}
	
	// MonoBehaviour
	void OnMasterServerEvent(MasterServerEvent msEvent) {
		if (msEvent == MasterServerEvent.HostListReceived) {
			hostList = MasterServer.PollHostList ();
		}
	}
	
	public void JoinServer (HostData hostData) {
		Network.Connect (hostData);
	}
	
	// MonoBehaviour
	void OnConnectedToServer () {
		SpawnPlayer ();
	}
	
	// MonoBehaviour
	void OnPlayerConnected (NetworkPlayer player) {
		SendGameMessage (player.ipAddress + " joined. " + NumberPlayers(true));
	}
	
	// MonoBehaviour
	void OnPlayerDisconnected (NetworkPlayer player) {
		Network.RemoveRPCs (player);
		Network.DestroyPlayerObjects (player);
		SendGameMessage (player.ipAddress + " left. " + NumberPlayers(false));
	}
	
	// MonoBehaviour
	void OnDisconnectedFromServer (NetworkDisconnection info) {
		if (Network.isServer) {
			SendGameMessage ("Local server connection lost.");
		} else {
			if (info == NetworkDisconnection.LostConnection) {
				SendGameMessage ("Lost connection to the server.");
			} else {
				SendGameMessage ("Disconnected from the server.");
			}
		}
		gameManager.RestartGame ();
	}
	
	// MonoBehaviour
	void OnFailedToConnect (NetworkConnectionError error) {
		SendGameMessage ("Could not connect to server.");
		SendGameMessage (error.ToString ());
		gameManager.ChangeObjectLight ("JoinIPButton", new Color (1.0f, 0.5f, 0.5f, 1.0f));
		
	}
	
	string NumberPlayers (bool add) {
		int plusOne = add ? 1 : 0;
		return (Network.connections.Length + plusOne) + "/" + Network.maxConnections;
	}
	
	public void SendGameMessage (string message) {
		chatList.Add (message);
		Debug.Log (FormatTime (Time.time) + " " + message);
		if (chatList.Count > 6) {
			chatList.RemoveAt (chatList.Count - 1);
		}
	}
	
	string FormatTime (float time) {
		int intTime = (int)time;
		int minutes = intTime / 60;
		int seconds = intTime % 60;
		int fraction = (int)time * 10;
		fraction = fraction % 10;
		
		var timeText = minutes.ToString("00") + ":";
		timeText += seconds.ToString("00") + ":";
		timeText += fraction.ToString("00");
		return timeText;
	}	
	
	public void setUI (bool active) {
		showUI = active;
	}
}

